core/src/main/scala/kafka/server/metadata/BrokerMetadataPublisher.scala

```diff
@@ -0,0 +1,249 @@
+          topicsDelta.image().getTopic(Topic.TRANSACTION_STATE_TOPIC_NAME).partitions().entrySet().forEach {
+            entry =>
+              if (entry.getValue().leader == brokerId) {
+                txnCoordinator.onResignation(entry.getKey(), Some(entry.getValue().leaderEpoch))
+              }
+          }
+        }
+        // If the transaction state topic changed in a way that's relevant to this broker,
+        // notify the transaction coordinator.
+        getTopicDelta(Topic.TRANSACTION_STATE_TOPIC_NAME, newImage, delta).foreach { topicDelta =>
+          topicDelta.newLocalLeaders(brokerId).forEach {
+            entry => txnCoordinator.onElection(entry.getKey(), entry.getValue().leaderEpoch)
+          }
+          topicDelta.newLocalFollowers(brokerId).forEach {
+            entry => txnCoordinator.onResignation(entry.getKey(), Some(entry.getValue().leaderEpoch))
+          }
+        }
+
+        // Notify the group coordinator about deleted topics.
+        val deletedTopicPartitions = new mutable.ArrayBuffer[TopicPartition]()
+        topicsDelta.deletedTopicIds().forEach { id =>
+          val topicImage = topicsDelta.image().getTopic(id)
+          topicImage.partitions().keySet().forEach {
+            id => deletedTopicPartitions += new TopicPartition(topicImage.name(), id)
+          }
+        }
+        if (!deletedTopicPartitions.isEmpty) {
+          groupCoordinator.handleDeletedPartitions(deletedTopicPartitions, RequestLocal.NoCaching)
+        }
+      }
+
+      // Apply configuration deltas.
+      Option(delta.configsDelta()).foreach { configsDelta =>
+        configsDelta.changes().keySet().forEach { configResource =>
+          val tag = configResource.`type`() match {
+            case ConfigResource.Type.TOPIC => Some(ConfigType.Topic)
+            case ConfigResource.Type.BROKER => Some(ConfigType.Broker)
+            case _ => None
+          }
+          tag.foreach { t =>
+            val newProperties = newImage.configs().configProperties(configResource)
+            dynamicConfigHandlers(t).processConfigChanges(configResource.name(), newProperties)
+          }
+        }
+      }
+
+      // Apply client quotas delta.
+      Option(delta.clientQuotasDelta()).foreach { clientQuotasDelta =>
+        clientQuotaMetadataManager.update(clientQuotasDelta)
+      }
+
+      if (_firstPublish) {
+        finishInitializingReplicaManager(newImage)
+      }
+    } catch {
+      case t: Throwable => error(s"Error publishing broker metadata at ${newHighestMetadataOffset}", t)
+        throw t
+    } finally {
+      _firstPublish = false
+    }
+  }
+
+  private def initializeManagers(): Unit = {
+    // Start log manager, which will perform (potentially lengthy)
+    // recovery-from-unclean-shutdown if required.
+    logManager.startup(metadataCache.getAllTopics())
+
+    // Start the replica manager.
+    replicaManager.startup()
```

Comment:
A little surprising to find this here given `BrokerServer` owns the rest of the lifecycle for these managers. Not sure if it would be any better with a callback to `BrokerServer`?

Reply:
I was thinking about that too. If I can find a way to condense the initialization into a single startup callback, I'll try to put the code back into `BrokerServer`. It still needs to happen in the context of the metadata listener thread, though.